test/conformance/labels_test.go

```diff
@@ -0,0 +1,130 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conformance
+
+import (
+	"testing"
+
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	serviceresourcenames "github.com/knative/serving/pkg/reconciler/v1alpha1/service/resources/names"
+	"github.com/knative/serving/test"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestLabelsPropagation(t *testing.T) {
+	clients := setup(t)
+
+	// Add test case specific name to its own logger.
+	logger := logging.GetContextLogger("TestLabelsPropagation")
+
+	var imagePath = test.ImagePath("helloworld")
+
+	var names test.ResourceNames
+	names.Service = test.AppendRandomString("pizzaplanet-service", logger)
+
+	defer tearDown(clients, names)
+	test.CleanupOnInterrupt(func() { tearDown(clients, names) }, logger)
+
+	logger.Info("Creating a new Service")
+	svc, err := test.CreateLatestService(logger, clients, names, imagePath)
+	if err != nil {
+		t.Fatalf("Failed to create Service: %v", err)
+	}
+	names.Route = serviceresourcenames.Route(svc)
+	names.Config = serviceresourcenames.Configuration(svc)
+
+	logger.Info("The Service will be updated with the name of the Revision once it is created")
+	var revisionName string
+	err = test.WaitForServiceState(clients.ServingClient, names.Service, func(s *v1alpha1.Service) (bool, error) {
+		if s.Status.LatestCreatedRevisionName != names.Revision {
+			revisionName = s.Status.LatestCreatedRevisionName
+			return true, nil
+		}
+		return false, nil
+	}, "ServiceUpdatedWithRevision")
+	revisionName, err = waitForServiceLatestCreatedRevision(clients, names)
+	if err != nil {
+		t.Fatalf("Service %s was not updated with the new revision: %v", names.Service, err)
+	}
+	names.Revision = revisionName
+
+	logger.Info("When the Service reports as Ready, everything should be ready.")
+	if err := test.WaitForServiceState(clients.ServingClient, names.Service, test.IsServiceReady, "ServiceIsReady"); err != nil {
+		t.Fatalf("The Service %s was not marked as Ready to serve traffic to Revision %s: %v", names.Service, names.Revision, err)
+	}
+
+	logger.Info("Validate Labels on Revision Object")
+	revision, err := clients.ServingClient.Revisions.Get(names.Revision, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error fetching Revision %s: %v", names.Revision, err)
+	}
+
+	if revConfigurationLabel, ok := revision.Labels["serving.knative.dev/configuration"]; ok {
+		if revConfigurationLabel != names.Config {
+			t.Fatalf("Expect confguration name in revision label %q but got %q ", names.Config, revConfigurationLabel)
+		}
+	} else {
+		t.Fatalf("Failed to get configuration name from Revision label")
+	}
```

Comment:
Totally optional: golang will return a empty value from a map lookup if you use the single-assignment form. Assuming that you know that `names.Config` is non-empty, you could write this as:

```golang
if revision.Labels["serving.knative.dev/configuration"] != names.Config {
  t.Fatalf(...)
}

Reply:
Also, I'd suggest using `t.Errorf()` here, to fail the test but allow it to keep executing. This will allow additional conditions to be tested and debugged at one time.

Reply:
This might also be an opportunity for a table-driven test; ala:

```golang
struct TestCase {
  Name string
  Fetcher func(string, metav1.Options) (interface, error)  // Must be an object 
  Labels map[string]string
}

expectedLabels := []TestCase{
  {
    Name: names.Revision,
    Fetcher: clients.ServingClient.Revisions.Get,
    Labels: {
        "serving.knative.dev/configuration": names.Config,
        "serving.knative.dev/service": names.Service,
    },
  },
  ...
}
```

Reply:
@evankanderson  
* I believe the table test would not work as your example above because Go doesn't allow [covariant return types](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Covariant_method_return_type) (golang/go#7512) as mentioned in a similar conversation in PR #2009
 * The `Fetcher` func can't return a generic interface, and the needed objects here do not confirm to any common interface.
 * The other option is to use [kubernetes dynamic client](https://github.com/kubernetes/client-go/tree/master/dynamic) to return `*unstructured.Unstructured` object and get the Labels from that. 
 * I feel this would be complicating this test too much since we have to change the client used, WDYT?